### PR TITLE
ENH: fully fleshed out Signal.set and EpicsSignal.set

### PR DESCRIPTION
**Pull request by klauer**
This will give us the option to use `set` similar to `Positioner`s for all signals, with asynchronous status completion objects.

* `Signal.set` uses `set_and_wait` (runs in a separate CAThread) - and so all signals derived from this will do the same
* `EpicsSignal.set` uses put completion if specified by the user (requires no background thread), and falls back to `set_and_wait`
* Switched Signal from using `StatusBase` -> `Status`

### CODE REVIEW

File: ophyd/signal.py [hunk 1]
```diff
@@ -97,7 +100,48 @@ def put(self, value, *, timestamp=None, force=False, **kwargs):
...
+        st : Status
+            This status object will be finished upon return in the
+            case of basic soft Signals
+        '''
+        def set_thread():
+            nonlocal timeout
+
+            if timeout is None:
+                timeout = 10
+                # TODO set_and_wait does not support a timeout of None
```
Comment (danielballan) [inline ophyd/signal.py]: Since `set_and_wait` polls and holds control in the meantime, I'm comfortable with this.

**Comment by danielballan** [comment]
Looks good to me. Thanks for tackling this -- a dragon at the bottom of a rabbit hole as Tom says. I'll wait for @tacaswell to sign off and merge.

File: ophyd/signal.py [hunk 2]
```diff
@@ -681,6 +728,45 @@ def put(self, value, force=False, **kwargs):
...
+        '''
+        if not self._put_complete:
+            return super().set(value, timeout=timeout, settle_time=settle_time)
+
+        # using put completion:
+        # timeout and settle time is handled by the status object.
+        st = Status(self, timeout=timeout, settle_time=settle_time)
+
+        def put_callback(**kwargs):
+            st._finished(success=True)
```
Comment (klauer) [inline ophyd/signal.py]: `settle_time` is ignored in this case - would need a background thread for it.

**Comment by tacaswell** [comment]
I would rather push the thread based `set` down to `EpicsSignalBase` and keep a base implementation in `Signal` that just return that it is done, but not going to push that (this is just my standard 'I want a sub-set of ophyd with out epics' line).

**Comment by klauer** [comment]
I think that having the `set_and_wait` implementation at the base of `Signal` makes sense, personally. 

You could make any sort of signal derived from that class, and `set_and_wait` would be called on it for `set()`. If you leave it as a no-operation (a status object that returns immediately), a delay between put() and get() returning it would not be captured.

**Pull request merged**